projects/opendr_ws/src/opendr_perception/scripts/fall_detection_node.py

```diff
@@ -96,76 +113,113 @@ def listen(self):
+        elif self.input_rgb_image_topic and not self.input_pose_topic:
+            rospy.loginfo("Fall detection node started in visualization mode.")
+        elif self.input_pose_topic and self.input_rgb_image_topic:
+            rospy.loginfo("Fall detection node started in both detection and visualization mode.")
+
         rospy.spin()
 
-    def callback(self, data):
+    def pose_callback(self, data):
+        """
+        Callback that processes the input pose data and publishes to the corresponding topics.
+        :param data: Input pose message
+        :type data: opendr_bridge.msg.OpenDRPose2D
+        """
+        if self.fall_performance_publisher:
+            start_time = perf_counter()
+        poses = [self.bridge.from_ros_pose(data)]
+        x, y, w, h = get_bbox(poses[0])  # Get bounding box for pose
+        detections = self.fall_detector.infer(poses)  # Run fall detection
+        fallen = detections[0][0].data  # Class: 1 = fallen, -1 = standing, 0 = can't detect
+
+        # Create Detection2D that contains the bbox of the pose as well as the detection class
+        ros_detection = self.bridge.to_ros_box(BoundingBox(left=x, top=y, width=w, height=h,
+                                                           name=fallen, score=poses[0].confidence))
+
+        if self.fall_performance_publisher:
+            end_time = perf_counter()
+            fps = 1.0 / (end_time - start_time)  # NOQA
+            fps_msg = Float32()
+            fps_msg.data = fps
+            self.fall_performance_publisher.publish(fps_msg)
+
+        self.fall_publisher.publish(ros_detection)
+
+    def image_callback(self, data):
         """
         Callback that processes the input data and publishes to the corresponding topics.
         :param data: Input image message
         :type data: sensor_msgs.msg.Image
         """
-        if self.performance_publisher:
+        if self.image_performance_publisher:
             start_time = perf_counter()
         # Convert sensor_msgs.msg.Image into OpenDR Image
         image = self.bridge.from_ros_image(data, encoding='bgr8')
 
         # Run fall detection
         detections = self.fall_detector.infer(image)
 
-        bboxes = BoundingBoxList([])
-        fallen_pose_id = 0
         for detection in detections:
             fallen = detection[0].data
+            pose = detection[1]
+            x, y, w, h = get_bbox(pose)
+
+            if self.image_performance_publisher:
+                end_time = perf_counter()
+                fps = 1.0 / (end_time - start_time)  # NOQA
+                fps_msg = Float32()
+                fps_msg.data = fps
+                self.image_performance_publisher.publish(fps_msg)
+
+            # Create Detection2D that contains the bbox of the pose as well as the detection class
+            self.fall_publisher.publish(self.bridge.to_ros_box(BoundingBox(left=x, top=y, width=w, height=h,
```

Comment:
You are right Stefania, thanks for noticing this. Basically i wanted to capture all the processing done before publishing in the performance calculation (in this case the `get_bbox` method). The same issue exists in the wave detection nodes also. 

Since the wave and fall detection work on each pose detected separately i went with calculating performance for each pose. Since every processed pose publishes the wave/fall result immediately, it is tricky to get the performance for all of the poses but exclude the time it takes to publish each result.

I can see two options: 
1. get performance from all the poses and publish one time, i.e. the performance of the whole callback, but excluding publishing, etc. I think it is tricky but doable, maybe the publishing gets all done after the loop in one go in a detection array, or manipulate the time values
2. keep it as it is but fix the issue you found, i.e. publish a performance result for each processed pose including the inference time or similar

What would you suggest?

Reply:
I pushed a fix according to our discussion, let me know how it looks so i can go ahead and apply it to the wave detection node as well as the ROS2 nodes.

Moreover, the same issue is present in other nodes that use a loop in their callback to do something after running inference, e.g. in face detection. I also identified several inconsistencies across all nodes in terms of how the performance is measured (in some places the bridge conversion is included, in others it is not). I will look into these issues and apply fixes in a separate PR to make it more manageable if that's ok with you.

Reply:
The new code seems fine!